framework/tlvf/src/include/tlvf/TlvList.h

```diff
@@ -0,0 +1,136 @@
+ */
+
+#ifndef _TlvList_H_
+#define _TlvList_H_
+
+#include "ieee_1905_1/eTlvType.h"
+#include <memory>
+#include <vector>
+#include <tlvf/BaseClass.h>
+
+namespace ieee1905_1 {
+
+class TlvList {
+
+public:
+    TlvList();
+    ~TlvList();
+
+public:
+    template <class T> std::shared_ptr<T> addClass()
+    {
+        std::shared_ptr<T> ptr;
+        if (m_buff) {
+            if (m_class_vector.size() == 0) {
+                ptr = std::make_shared<T>(m_buff, m_parse, m_swap);
+            } else {
+                ptr = std::make_shared<T>(m_class_vector.back(), m_parse, m_swap);
+            }
+            if (!ptr || ptr->isInitialized() == false) {
+                return std::shared_ptr<T>();
+            } else {
+                m_class_vector.push_back(std::shared_ptr<BaseClass>(ptr));
+            }
+        } else {
+            return std::shared_ptr<T>();
+        }
+        return ptr;
+    }
+
+    template <class T> std::shared_ptr<T> dynamicCast(std::shared_ptr<BaseClass> ptr) const
+    {
+        return std::dynamic_pointer_cast<T>(ptr);
+    }
+
+    /**
+     * @brief Get the Class object at index idx in the all classes array
+     *
+     * @param idx index in the all classes array
+     * @return std::shared_ptr<BaseClass> to the class object at index idx, nullptr if not found
+     */
+    std::shared_ptr<BaseClass> getClass(size_t idx) const;
+
+    /**
+     * @brief Get the (first) Class object
+     *
+     * @tparam T class template
+     * @return std::shared_ptr<T> to the first object found of type T, nullptr if not found
+     */
+    template <class T> std::shared_ptr<T> getClass() const
+    {
+        for (size_t idx = 0; idx < getClassCount(); idx++) {
+            if (auto c = std::dynamic_pointer_cast<T>(getClass(idx)))
+                return c;
+        }
+        return nullptr;
+    }
+
+    /**
+     * @brief Get a class object of type T in index `idx` in the logical array containing
+     *        all classes of type T.
+     *
+     * @tparam T class template
+     * @param idx index in the class T array
+     * @return std::shared_ptr<T> to the T class at index `idx` in the class T array
+     */
+    template <class T> std::shared_ptr<T> getClass(size_t idx) const
+    {
+        size_t idx_ = 0;
+        for (size_t i = 0; i < getClassCount(); i++) {
+            if (auto c = std::dynamic_pointer_cast<T>(getClass(i))) {
+                if (idx_++ == idx)
+                    return c;
+            }
+        }
+        return nullptr;
+    }
+
+    /**
+     * @brief Get the number of classes of type T
+     *
+     * @tparam T class template
+     * @return size_t number of classes of type T
+     */
+    template <class T> size_t getClassCount() const
+    {
+        size_t count = 0;
+        for (size_t i = 0; i < getClassCount(); i++) {
+            if (auto c = std::dynamic_pointer_cast<T>(getClass(i))) {
+                count++;
+            }
+        }
+        return count;
+    }
```

Comment:
I think this is only used for a pattern like this:
```
for (int idx = 0; idx < t.getClassCount<Foo>(); idx++) {
   auto foo = t.getClass<Foo>(idx);
   ...
}
```

It would be a lot better to convert this to an iterator pattern, but I guess that's a lot more complicated.

Reply:
It would, and it is, so I'll leave it as is for now :)